services/brig/src/Brig/API/Public.hs

```diff
@@ -1220,26 +1224,36 @@ getUserDisplayNameH (_ ::: self) = do
     Just n -> json $ object ["name" .= n]
     Nothing -> setStatus status404 empty
 
-listUsersH :: JSON ::: UserId ::: Either (List UserId) (Range 1 4 (List Handle)) -> Handler Response
-listUsersH (_ ::: self ::: qry) =
-  toResponse <$> listUsers self qry
-  where
-    toResponse = \case
-      [] -> setStatus status404 empty
-      ps -> json ps
-
--- | 'listUsers' only handles listing local users by ID or handle. We decided to
--- create a new federation aware endpoint which accepts federation aware Ids or
--- handles in the request body using a 'POST' request to avoid specifying
--- complex objects in the query parameters.
-listUsers :: UserId -> Either (List UserId) (Range 1 4 (List Handle)) -> Handler [Public.UserProfile]
-listUsers self = \case
-  Left us -> do
-    domain <- viewFederationDomain
-    byIds $ map (`Qualified` domain) (fromList us)
-  Right hs -> do
-    us <- getIds (fromList $ fromRange hs)
-    filterHandleResults self =<< byIds us
+-- FUTUREWORK: Make servant understand that at least one of these is required
+listUsersByUnqualifiedIdsOrHandles :: UserId -> Maybe (CommaSeparatedList UserId) -> Maybe (Range 1 4 (CommaSeparatedList Handle)) -> Handler [Public.UserProfile]
+listUsersByUnqualifiedIdsOrHandles self mUids mHandles = do
+  domain <- viewFederationDomain
+  case (mUids, mHandles) of
+    (Just uids, _) -> listUsersByIdsOrHandles self (Public.ListUsersByIds ((`Qualified` domain) <$> fromCommaSeparatedList uids))
+    (_, Just handles) ->
+      let normalRangedList = fromCommaSeparatedList $ fromRange handles
+          qualifiedList = (`Qualified` domain) <$> normalRangedList
+          -- Use of unsafeRange here is ok only because we know that 'handles'
+          -- is valid for 'Range 1 4'. However, we must not forget to keep this
+          -- annotation here otherwise a change in 'Public.ListUsersByHandles'
+          -- could cause this code to break.
+          qualifiedRangedList :: Range 1 4 [Qualified Handle] = unsafeRange qualifiedList
+       in listUsersByIdsOrHandles self (Public.ListUsersByHandles qualifiedRangedList)
+    (Nothing, Nothing) -> throwStd $ badRequest "at least one ids or handles must be provided"
```

Comment:
So in case both are provided, we just silently ignore the provided handles? Would be nicer to fail, maybe reducing the boilderplate by using some kind of combinator `Maybe a -> Maybe b -> Handler (Either a b)` or even API combinator.

Reply:
I decided to not do this for backwards compatibility. Currently nothing breaks if a client sends both of these and handles get ignored if uids are present. So, maybe it is ok to live with this in the old endpoint.
That being said I agree that it would be nicer to throw an error and since we're making a new endpoint, I will just make the `parseJSON` for the input fail if both are present.

Reply:
Ah, you're right. Didn't know it behaved that way, but agree that it's best to keep it for the deprecated endpoint.